app/health_check.go

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package app
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/mattermost/mattermost-server/mlog"
+	"github.com/mattermost/mattermost-server/model"
+)
+
+type PluginHealthCheckJob struct {
+	server    *Server
+	task      *model.ScheduledTask
+	taskMutex sync.Mutex
+}
+
+const (
+	PLUGIN_HEALTH_CHECK_TASK_NAME = "Plugin Health Check"
+	PLUGIN_HEALTH_CHECK_INTERVAL  = 5 // seconds
+	HEALTH_CHECK_FAIL_LIMIT       = 3
+	ENABLE_HEALTH_CHECK_JOB       = false
+)
+
+func (s *Server) InitPluginHealthCheckJob() {
+	if ENABLE_HEALTH_CHECK_JOB {
+		s.PluginsLock.RLock()
+		pluginsEnvironment := s.PluginsEnvironment
+		s.PluginsLock.RUnlock()
+
+		if pluginsEnvironment == nil {
+			return
+		}
+
+		if s.PluginHealthCheck == nil {
+			s.PluginHealthCheck = NewPluginHealthCheckJob(s)
+		}
+
+		s.PluginHealthCheck.Start()
+	}
+}
+
+func (job *PluginHealthCheckJob) Start() {
+	mlog.Debug(fmt.Sprintf("Plugin health check job starting. Sending health check pings every %v seconds.", PLUGIN_HEALTH_CHECK_INTERVAL))
+	newTask := model.CreateRecurringTask(PLUGIN_HEALTH_CHECK_TASK_NAME, job.RunPluginHealthChecks, time.Duration(PLUGIN_HEALTH_CHECK_INTERVAL)*time.Second)
+	job.taskMutex.Lock()
+	oldTask := job.task
+	job.task = newTask
+	job.taskMutex.Unlock()
+
+	if oldTask != nil {
+		oldTask.Cancel()
+	}
+}
+
+func (job *PluginHealthCheckJob) RunPluginHealthChecks() {
+	if pluginsEnvironment := job.server.PluginsEnvironment; pluginsEnvironment != nil {
+		for _, ap := range pluginsEnvironment.Active() {
+			job.RunPluginHealthCheck(ap)
+		}
+	}
+}
+
+func (job *PluginHealthCheckJob) RunPluginHealthCheck(plugin *model.BundleInfo) {
+	if pluginsEnvironment := job.server.PluginsEnvironment; pluginsEnvironment != nil {
+		id := plugin.Manifest.Id
+
+		if plugin.HealthCheckFails == HEALTH_CHECK_FAIL_LIMIT {
+			mlog.Debug(fmt.Sprintf("Plugin `%v` exceeded health check failure threshold. Restarting Plugin.", id))
+			pluginsEnvironment.RestartPlugin(id)
```

Comment:
If I understand the logic correct, this will ensure the plugin get restarted every time it has failed 3 time? Do we really want to restart it _forever_? Maybe just add a restart limit instead?

Reply:
THAT ^^. There needs to be a policy for retrying, and a policy for restarting, both configurable. The loop of restart/crash/detect must be avoided.

Reply:
@levb I agree. This should be a main focus of this monitoring/restart effort. Once that loop has begun, it will happen forever as @hanzei pointed out. As such, I'm treating this specific issue with very high importance for design moving forward.

Reply:
My thoughts on the policy are:

For me the three cases (Ping failed, health check failed, process not found) all require separate policies.

1. Ping failed: Retry ping A times, then restart
2. Health check: Restart immediately but only B times in C minutes. If the limit B is reached, disable the plugin.
2. Health check: Same as 2. but with a different limit.

Thoughts on this?

Reply:
I have the different failure cases implemented with a common shared fail limit at the moment, but I think they could benefit by treating them as separate policies as you mention. The reasons for failure are different, and shouldn't necessarily be handled the same way. I'd be curious to hear others' thoughts on this.